packages/ddp-server/livedata_server.js

```diff
@@ -423,52 +434,61 @@ Object.assign(Session.prototype, {
     // Destroy this session, even if it's not registered at the
     // server. Stop all processing and tear everything down. If a socket
     // was attached, close it.
-
-    // Already destroyed.
-    if (! self.inQueue)
+    if (self._removeTimeoutHandle)
       return;
 
-    // Drop the merge box data immediately.
-    self.inQueue = null;
-    self.collectionViews = new Map();
-
-    if (self.heartbeat) {
-      self.heartbeat.stop();
-      self.heartbeat = null;
-    }
-
     if (self.socket) {
       self.socket.close();
       self.socket._meteorSession = null;
     }
+    self.server._removeSession(self, () => {
 
-    Package['facts-base'] && Package['facts-base'].Facts.incrementServerFact(
-      "livedata", "sessions", -1);
-
-    Meteor.defer(function () {
-      // stop callbacks can yield, so we defer this on close.
-      // sub._isDeactivated() detects that we set inQueue to null and
-      // treats it as semi-deactivated (it will ignore incoming callbacks, etc).
-      self._deactivateAllSubscriptions();
-
-      // Defer calling the close callbacks, so that the caller closing
-      // the session isn't waiting for all the callbacks to complete.
-      _.each(self._closeCallbacks, function (callback) {
-        callback();
+      Package['facts-base'] && Package['facts-base'].Facts.incrementServerFact(
+        "livedata", "sessions", -1);
+
+      self.inQueue = null;
+      self.collectionViews = new Map();
+
+      if (self.heartbeat) {
+        self.heartbeat.stop();
+        self.heartbeat = null;
+      }
+
+      Meteor.defer(function () {
+        // stop callbacks can yield, so we defer this on close.
+        // sub._isDeactivated() detects that we set inQueue to null and
+        // treats it as semi-deactivated (it will ignore incoming callbacks, etc).
+        self._deactivateAllSubscriptions();
+
+        // Defer calling the close callbacks, so that the caller closing
+        // the session isn't waiting for all the callbacks to complete.
+        _.each(self._closeCallbacks, function (callback) {
+          callback();
+        });
       });
     });
 
-    // Unregister the session.
-    self.server._removeSession(self);
   },
 
   // Send a message (doing nothing if no socket is connected right now.)
   // It should be a JSON object (it will be stringified.)
   send: function (msg) {
     var self = this;
+    const isPingPong = ['ping', 'pong'].includes(msg.msg);
+    if (self.messageQueue && !isPingPong) {
+      self.messageQueue.push(msg);
+      if (self.messageQueue.length > MAX_QUEUE_LENGTH) {
```

Comment:
max queue length should be also an option in the package to be replaced if someone wants to customize.